src/http-api/errors.ts

```diff
@@ -63,11 +67,66 @@ export class MatrixError extends HTTPError {
         if (url) {
             message = `${message} (${url})`;
         }
-        super(`MatrixError: ${message}`, httpStatus);
+        super(`MatrixError: ${message}`, httpStatus, httpHeaders);
         this.errcode = errorJson.errcode;
         this.name = errorJson.errcode || "Unknown error code";
         this.data = errorJson;
     }
+
+    /**
+     * @returns whether this error is due to rate-limiting.
+     */
+    public isRateLimitError(): boolean {
+        return this.errcode === "M_LIMIT_EXCEEDED" || (this.errcode === "M_UNKNOWN" && this.httpStatus === 429);
+    }
+
+    /**
+     * @returns the recommended delay in milliseconds to wait before retrying
+     * the request that triggered this error, or null if no delay is recommended.
+     * @throws Error if the recommended delay is an invalid value.
+     */
+    public getRetryAfterMs(): number | null {
+        const retryAfter = this.httpHeaders?.get("Retry-After");
+        if (retryAfter != null) {
+            if (/^\d+$/.test(retryAfter)) {
+                const ms = Number.parseInt(retryAfter) * 1000;
+                if (!Number.isFinite(ms)) {
+                    throw new Error("Retry-After header integer value is too large");
+                }
+                return ms;
+            }
+            const date = new Date(retryAfter);
+            if (date.toUTCString() !== retryAfter) {
+                throw new Error("Retry-After header value is not a valid HTTP-date or non-negative decimal integer");
+            }
+            return date.getTime() - Date.now();
+        }
+        // Note: retry_after_ms is deprecated as of spec version v1.10
+        if (this.errcode === "M_LIMIT_EXCEEDED" && "retry_after_ms" in this.data) {
+            if (!Number.isInteger(this.data.retry_after_ms)) {
+                throw new Error("retry_after_ms is not an integer");
+            }
+            return this.data.retry_after_ms;
+        }
+        return null;
+    }
+}
+
+/**
+ * @returns the recommended delay in milliseconds to wait before retrying
+ * the request that triggered {@link error}, or {@link defaultMs} if no valid
+ * delay is recommended.
+ */
+export function safeGetRetryAfterMs(error: unknown, defaultMs: number): number {
+    if (!(error instanceof MatrixError) || !error.isRateLimitError()) {
+        return defaultMs;
+    }
```

Comment:
Why is the check on `isRateLimitError` necessary? If the error is not due to rate-limiting, then we would expect `error.getRetryAfterMs()` to return null, and therefore we fall back to `defaultMs` anyway

Reply:
Good point.  I had done this because of the rule to give the errcode precedence over the HTTP status code, but as long as a `Retry-Header` value is there & callers want to apply _some_ delay regardless of the errcode, they may as well use the header value if one is present.